llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,128 @@
+
+#include "llvm/SYCLLowerIR/SYCLConditionalCallOnDevice.h"
+
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Intrinsics.h"
+
+using namespace llvm;
+
+cl::opt<std::string>
+    UniquePrefixOpt("sycl-conditional-call-on-device-unique-prefix",
+                    cl::Optional, cl::Hidden,
+                    cl::desc("Set unique prefix for a translation unit, "
+                             "required for funtions with external linkage"),
+                    cl::init(""));
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+
+    if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+      continue;
+
+    if (F.hasFnAttribute("sycl-call-if-on-device-conditionally") &&
+        ("true" == F.getFnAttribute("sycl-call-if-on-device-conditionally")
+                       .getValueAsString()))
+      FCallers.push_back(&F);
+  }
+
+  // A vector instead of DenseMap to make LIT tests predictable
+  SmallVector<std::pair<Function *, Function *>, 8> FCallersToFActions;
+  for (Function *FCaller : FCallers) {
+    // Find call to @CallableXXX in call_if_on_device_conditionally function
+    // (FAction). FAction should be a literal (i.e. not a pointer). The
+    // structure of the header file ensures that there is exactly one such
+    // instruction.
+    for (Instruction &I : instructions(FCaller)) {
+      if (auto *CI = dyn_cast<CallInst>(&I);
+          CI && (Intrinsic::IndependentIntrinsics::not_intrinsic ==
+                 CI->getIntrinsicID())) {
+        FCallersToFActions.push_back(
+            std::make_pair(FCaller, CI->getCalledFunction()));
+        break;
+      }
+    }
+  }
+
+  int FCallerIndex = 1;
+  for (const auto &FCallerToFAction : FCallersToFActions) {
+    Function *FCaller = FCallerToFAction.first;
+    Function *FAction = FCallerToFAction.second;
+
+    // Create a new function type with an additional function pointer argument
+    SmallVector<Type *, 4> NewParamTypes;
+    Type *FActionType = FAction->getType();
+    NewParamTypes.push_back(
+        PointerType::getUnqual(FActionType)); // Add function pointer to FAction
+    FunctionType *OldFCallerType = FCaller->getFunctionType();
+    for (Type *Ty : OldFCallerType->params())
+      NewParamTypes.push_back(Ty);
+
+    auto *NewFCallerType =
+        FunctionType::get(OldFCallerType->getReturnType(), NewParamTypes,
+                          OldFCallerType->isVarArg());
+
+    // Create a new function with the updated type and rename it to
+    // call_if_on_device_conditionally_GUID_N
+    if (!UniquePrefixOpt.empty())
+      UniquePrefix = UniquePrefixOpt;
+    Twine NewFCallerName = Twine(FCaller->getName()) + "_" + UniquePrefix +
```

Comment:
`Twine` documentation explicitly says that it shouldn't be stored, but it is not clear to me whether it means in-memory (i.e. in a container), of if it even includes a regular stack (i.e. it should only be a temporary passed as an argument). However, I don't have enough knowledge about it to say for sure if this use is safe or not.

Reply:
Yeah, I see many usages in LLVM codebase where `Twine` variable is stored (doesn't mean that it's correct), and I moved the concatenation to the argument itself now just to follow this limitation.